File: app/agent/generic_agent.py
"""
Generic Agent implementation that uses a Working Backwards methodology.
This agent starts from the goal state and works backwards to create a plan,
then executes the plan step by step using available tools.
"""

from typing import Dict, List, Optional, Any

from pydantic import Field, model_validator

from app.agent.toolcall import ToolCallAgent
from app.logger import logger
from app.prompt.generic_agent import (
    SYSTEM_PROMPT,
    NEXT_STEP_PROMPT,
    PLANNING_TEMPLATE,
    EXECUTION_STATUS_TEMPLATE
)
from app.schema import Message
from app.tool import ToolCollection, Terminate


class GenericAgent(ToolCallAgent):
    """
    An agent that implements the Working Backwards methodology to solve problems.

    This agent first clarifies the goal, then uses step-back questioning to identify
    prerequisites for each step, working backwards until reaching actions that can be
    taken from the current state. It then reorders these steps and executes them.
    """

    name: str = "generic_agent"
    description: str = "A general-purpose agent that uses Working Backwards methodology"

    system_prompt: str = SYSTEM_PROMPT
    next_step_prompt: str = NEXT_STEP_PROMPT

    # Tracking goal and plan
    goal_state: str = Field(default="")
    current_state: str = Field(default="")
    current_progress: str = Field(default="No progress yet. Planning phase.")

    # Working backwards tracking
    backwards_steps: List[Dict[str, Any]] = Field(default_factory=list)
    forward_plan: List[Dict[str, Any]] = Field(default_factory=list)

    # Execution tracking
    plan_ready: bool = Field(default=False)
    current_step_index: int = Field(default=0)
    completed_steps: List[Dict[str, Any]] = Field(default_factory=list)
    pending_steps: List[Dict[str, Any]] = Field(default_factory=list)

    # Additional state for flexibility
    state_variables: Dict[str, Any] = Field(default_factory=dict)

    max_steps: int = 30

    def __init__(self, **data):
        """Initialize the GenericAgent with custom tools if provided."""
        # If no tools are provided, add a basic set
        if "available_tools" not in data:
            data["available_tools"] = ToolCollection([Terminate()])

        # Ensure Terminate tool is always available
        terminate_tool_exists = False
        for tool in data.get("available_tools", []):
            if (isinstance(tool, Terminate) or
                    (hasattr(tool, "name") and tool.name == "terminate")):
                terminate_tool_exists = True
                break

        if not terminate_tool_exists and "available_tools" in data:
            data["available_tools"].add_tool(Terminate())

        super().__init__(**data)

    @model_validator(mode="after")
    def validate_agent_state(self) -> "GenericAgent":
        """Validate the agent's state and ensure all required fields are initialized."""
        if not self.goal_state:
            self.goal_state = "Not yet specified"

        if not self.current_state:
            self.current_state = "Initial state"

        # Ensure special tools are properly tracked
        terminate_names = ["terminate", "finish", "complete"]
        self.special_tool_names = list(set(self.special_tool_names + [
            tool for tool in terminate_names
            if tool in [t.name for t in self.available_tools.tools]
        ]))

        return self

    async def set_goal(self, goal: str) -> str:
        """
        Set the agent's goal state.

        Args:
            goal: The goal state the agent should work towards.

        Returns:
            A confirmation message.
        """
        self.goal_state = goal
        logger.info(f"Goal set: {goal}")

        # Clear previous plan if it exists
        self.backwards_steps = []
        self.forward_plan = []
        self.plan_ready = False
        self.current_step_index = 0
        self.completed_steps = []
        self.pending_steps = []

        # Add goal setting message to memory
        self.update_memory("system", f"Goal set: {goal}")

        # Generate initial clarification of the goal
        goal_clarification_msg = Message.user_message(
            f"My goal is: {goal}\n\nPlease clarify this goal in concrete, "
            f"specific terms and begin the Working Backwards analysis."
        )
        self.memory.add_message(goal_clarification_msg)

        return f"Goal set to: {goal}. I will now analyze this goal and work backwards."

    async def think(self) -> bool:
        """
        Process the current state and decide the next action.

        Returns:
            bool: True if there are actions to take, False otherwise.
        """
        formatted_prompt = NEXT_STEP_PROMPT.format(
            goal_state=self.goal_state,
            current_progress=self.current_progress,
            current_state=self.current_state
        )

        # Add next step prompt to memory
        self.messages.append(Message.user_message(formatted_prompt))

        # Get response with tool options
        result = await super().think()

        # If we're done planning but haven't started executing, organize the plan
        if not self.plan_ready and self.backwards_steps and not self.tool_calls:
            await self._organize_plan()

        return result

    async def act(self) -> str:
        """
        Execute decided actions and update plan progress.

        Returns:
            str: Result of the action.
        """
        # Execute the action
        result = await super().act()

        # Update plan status if we have a plan
        if self.plan_ready and self.forward_plan:
            # Find the current step and mark it as completed
            if self.current_step_index < len(self.forward_plan):
                completed_step = self.forward_plan[self.current_step_index].copy()
                completed_step["result"] = result
                self.completed_steps.append(completed_step)

                # Update progress tracking
                self.current_progress = (
                    f"Completed {len(self.completed_steps)}/{len(self.forward_plan)} "
                    f"steps. Last completed: {completed_step.get('description', 'Step')}"
                )

                # Move to next step
                self.current_step_index += 1

                # Update pending steps
                if self.current_step_index < len(self.forward_plan):
                    self.pending_steps = self.forward_plan[self.current_step_index:]
                else:
                    self.pending_steps = []

                logger.info(
                    f"Completed step {self.current_step_index}/{len(self.forward_plan)}:"
                    f" {result[:100]}..."
                )

        return result

    async def run(self, request: Optional[str] = None) -> str:
        """
        Execute the agent's main workflow.

        Args:
            request: Initial request or goal for the agent.

        Returns:
            str: Summary of execution results.
        """
        if request:
            await self.set_goal(request)

        return await super().run()

    async def _organize_plan(self) -> None:
        """
        Organize the backwards steps into a forward execution plan.
        This reverses the order of steps and prepares them for execution.
        """
        if not self.backwards_steps:
            return

        # Prepare to create the forward plan using LLM
        steps_analysis = "\n".join([
            f"{i+1}. {step.get('description', 'Step')}"
            for i, step in enumerate(self.backwards_steps)
        ])

        planning_prompt = PLANNING_TEMPLATE.format(
            goal=self.goal_state,
            current_state=self.current_state,
            backwards_analysis=steps_analysis,
            forward_plan="To be determined",
            tools_required=", ".join([t.name for t in self.available_tools.tools]),
            success_criteria="To be determined",
            potential_challenges="To be determined",
            monitoring_approach="To be determined"
        )

        # Ask LLM to create the forward plan
        messages = self.messages + [Message.user_message(planning_prompt)]
        response = await self.llm.ask(messages=messages)

        # Parse the response to extract the forward plan
        # This is simplified and might need more robust parsing in a real implementation
        self.update_memory("assistant", response)

        # Create simple forward plan by reversing backwards steps
        self.forward_plan = list(reversed(self.backwards_steps))

        # Mark the plan as ready
        self.plan_ready = True
        self.pending_steps = self.forward_plan.copy()

        # Log the plan creation
        plan_summary = "\n".join([
            f"{i+1}. {step.get('description', 'Step')}"
            for i, step in enumerate(self.forward_plan)
        ])

        logger.info(
            f"Forward execution plan created with {len(self.forward_plan)} steps:\n"
            f"{plan_summary}"
        )

        # Add the plan to memory
        self.update_memory(
            "system",
            f"Forward execution plan created with {len(self.forward_plan)} steps:\n"
            f"{plan_summary}"
        )

    def _update_execution_status(self) -> str:
        """
        Create a formatted status of the execution plan.

        Returns:
            str: Formatted execution status.
        """
        if not self.plan_ready:
            return "Plan is still being formulated."

        plan_steps = "\n".join([
            f"{i+1}. {step.get('description', 'Step')}"
            for i, step in enumerate(self.forward_plan)
        ])

        completed_steps = "\n".join([
            f"✓ {i+1}. {step.get('description', 'Step')}"
            for i, step in enumerate(self.completed_steps)
        ]) if self.completed_steps else "None yet."

        pending_steps = "\n".join([
            f"○ {self.current_step_index+i+1}. {step.get('description', 'Step')}"
            for i, step in enumerate(self.pending_steps)
        ]) if self.pending_steps else "None remaining."

        if self.current_step_index < len(self.forward_plan):
            current_step = (
                f"Currently executing step {self.current_step_index+1}/"
                f"{len(self.forward_plan)}: "
                f"{self.forward_plan[self.current_step_index].get('description', 'Step')}"
            )
        else:
            current_step = "All steps completed."

        return EXECUTION_STATUS_TEMPLATE.format(
            goal=self.goal_state,
            plan_steps=plan_steps,
            current_step=current_step,
            completed_steps=completed_steps,
            pending_steps=pending_steps,
            observations=self.current_progress,
            adjustments="None required at this time."
        )

    def add_backward_step(
        self, description: str, tools_needed: List[str], prerequisites: List[str]
    ) -> None:
        """
        Add a step to the backwards planning process.

        Args:
            description: Description of this step
            tools_needed: Tools required for this step
            prerequisites: What must be true before this step
        """
        self.backwards_steps.append({
            "description": description,
            "tools_needed": tools_needed,
            "prerequisites": prerequisites
        })

    async def get_execution_status(self) -> str:
        """
        Get the current execution status.

        Returns:
            str: Formatted execution status.
        """
        return self._update_execution_status()

File: src/app/agent/generic_agent.ts
/**
 * Generic Agent Implementation
 * 
 * This module implements a generic agent that uses the Working Backwards methodology
 * to solve problems step by step.
 */

export interface Tool {
  name: string;
  description: string;
  execute: (params: any) => Promise<any>;
}

export interface Step {
  id: string;
  description: string;
  prerequisites: string[];
  toolsNeeded: string[];
  completed: boolean;
  result?: any;
}

export interface AgentState {
  goalState: string;
  currentState: string;
  backwardsSteps: Step[];
  forwardPlan: Step[];
  completedSteps: Step[];
  pendingSteps: Step[];
  currentStepIndex: number;
  planReady: boolean;
  currentProgress: string;
}

/**
 * GenericAgent class implementing the Working Backwards methodology
 * 
 * Formula: F(Generic Agentive Workflow) = 
 *   Goal → Step-back questions (Z→Y→X→...→A) →
 *   Reordering (A→B→C→...→Z) →
 *   ∫(Each step with reasoning + tool use)d(step) →
 *   Integrated execution plan (Result)
 */
export class GenericAgent {
  private state: AgentState;
  private tools: Map<string, Tool>;
  private maxSteps: number;
  private thinking: boolean;
  
  /**
   * Create a new GenericAgent
   * 
   * @param goalState The desired end state to achieve
   * @param tools Available tools for the agent to use
   * @param maxSteps Maximum number of steps to execute
   */
  private listeners: Map<string, Array<(data: any) => void>>;
  
  constructor(goalState: string, tools: Tool[] = [], maxSteps: number = 10) {
    this.tools = new Map(tools.map(tool => [tool.name, tool]));
    this.maxSteps = maxSteps;
    this.thinking = false;
    this.listeners = new Map();
    
    this.state = {
      goalState,
      currentState: "Initial state",
      backwardsSteps: [],
      forwardPlan: [],
      completedSteps: [],
      pendingSteps: [],
      currentStepIndex: 0,
      planReady: false,
      currentProgress: "Planning phase - Working backwards from goal"
    };
  }
  
  /**
   * Register an event listener
   * 
   * @param event The event to listen for
   * @param listener The listener function
   */
  on(event: string, listener: (data: any) => void): void {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, []);
    }
    this.listeners.get(event)?.push(listener);
  }
  
  /**
   * Emit an event
   * 
   * @param event The event to emit
   * @param data The data to pass to listeners
   */
  private emit(event: string, data: any): void {
    const eventListeners = this.listeners.get(event);
    if (eventListeners) {
      for (const listener of eventListeners) {
        listener(data);
      }
    }
  }
  
  /**
   * Start the Working Backwards process
   * 
   * @returns Promise that resolves when planning is complete
   */
  async planBackwards(): Promise<void> {
    this.thinking = true;
    this.emit('thinking', { message: `Planning backwards from goal: ${this.state.goalState}` });
    
    try {
      await this.stepBackQuestions();
      
      this.reorderSteps();
      
      this.state.planReady = true;
      this.state.pendingSteps = [...this.state.forwardPlan];
      this.thinking = false;
      
      this.emit('plan-ready', { 
        message: "Plan is ready for execution",
        plan: this.state.forwardPlan
      });
    } catch (error) {
      this.thinking = false;
      this.emit('error', { message: `Planning error: ${error}` });
      throw error;
    }
  }
  
  /**
   * Execute the forward plan
   * 
   * @returns Promise that resolves when execution is complete
   */
  async execute(): Promise<void> {
    if (!this.state.planReady) {
      throw new Error("Cannot execute before planning is complete");
    }
    
    this.emit('execution-start', { 
      message: "Starting execution of plan",
      plan: this.state.forwardPlan
    });
    
    try {
      while (this.state.currentStepIndex < this.state.forwardPlan.length) {
        if (this.state.currentStepIndex >= this.maxSteps) {
          this.emit('max-steps-reached', { 
            message: `Reached maximum steps (${this.maxSteps})`,
            completedSteps: this.state.completedSteps,
            remainingSteps: this.state.pendingSteps
          });
          break;
        }
        
        const currentStep = this.state.forwardPlan[this.state.currentStepIndex];
        
        this.emit('step-start', { 
          message: `Executing step ${this.state.currentStepIndex + 1}/${this.state.forwardPlan.length}`,
          step: currentStep
        });
        
        const result = await this.executeStep(currentStep);
        
        currentStep.completed = true;
        currentStep.result = result;
        
        this.state.completedSteps.push(currentStep);
        this.state.currentStepIndex++;
        this.state.pendingSteps = this.state.forwardPlan.slice(this.state.currentStepIndex);
        
        this.state.currentProgress = `Completed ${this.state.completedSteps.length}/${this.state.forwardPlan.length} steps`;
        
        this.emit('step-complete', { 
          message: `Completed step ${this.state.currentStepIndex}/${this.state.forwardPlan.length}`,
          step: currentStep,
          result
        });
      }
      
      this.emit('execution-complete', { 
        message: "Execution complete",
        completedSteps: this.state.completedSteps
      });
    } catch (error) {
      this.emit('error', { message: `Execution error: ${error}` });
      throw error;
    }
  }
  
  /**
   * Run the full agent workflow (plan and execute)
   * 
   * @returns Promise that resolves when the workflow is complete
   */
  async run(): Promise<any> {
    await this.planBackwards();
    await this.execute();
    return this.getResults();
  }
  
  /**
   * Get the current results of the agent's work
   * 
   * @returns Object containing the agent's results
   */
  getResults(): any {
    return {
      goalState: this.state.goalState,
      completedSteps: this.state.completedSteps,
      pendingSteps: this.state.pendingSteps,
      currentProgress: this.state.currentProgress,
      success: this.state.currentStepIndex >= this.state.forwardPlan.length
    };
  }
  
  /**
   * Get the current state of the agent
   * 
   * @returns The agent's current state
   */
  getState(): AgentState {
    return { ...this.state };
  }
  
  /**
   * Add a tool to the agent's toolkit
   * 
   * @param tool The tool to add
   */
  addTool(tool: Tool): void {
    this.tools.set(tool.name, tool);
  }
  
  /**
   * Remove a tool from the agent's toolkit
   * 
   * @param toolName The name of the tool to remove
   */
  removeTool(toolName: string): void {
    this.tools.delete(toolName);
  }
  
  /**
   * Add a backward step manually
   * 
   * @param description Step description
   * @param prerequisites Prerequisites for this step
   * @param toolsNeeded Tools needed for this step
   */
  addBackwardStep(description: string, prerequisites: string[] = [], toolsNeeded: string[] = []): void {
    const step: Step = {
      id: `back-${Date.now()}-${this.state.backwardsSteps.length}`,
      description,
      prerequisites,
      toolsNeeded,
      completed: false
    };
    
    this.state.backwardsSteps.push(step);
    this.emit('backward-step-added', { step });
  }
  
  /**
   * Private method to simulate the step-back questioning process
   * This would typically involve LLM calls in a real implementation
   */
  private async stepBackQuestions(): Promise<void> {
    
    let currentState = this.state.goalState;
    
    const numSteps = Math.floor(Math.random() * 3) + 3;
    
    for (let i = 0; i < numSteps; i++) {
      await new Promise(resolve => setTimeout(resolve, 500)); // Simulate thinking time
      
      const stepDescription = `Step to achieve: ${currentState}`;
      const prerequisites = [`Prerequisite ${i+1} for ${currentState}`];
      const toolsNeeded = Array.from(this.tools.keys()).slice(0, Math.floor(Math.random() * this.tools.size));
      
      this.addBackwardStep(stepDescription, prerequisites, toolsNeeded);
      
      currentState = `Precondition ${i+1} for ${this.state.goalState}`;
      
      this.emit('thinking-progress', { 
        message: `Working backwards: Step ${i+1}`,
        currentState
      });
    }
  }
  
  /**
   * Private method to reorder backward steps into a forward plan
   */
  private reorderSteps(): void {
    this.state.forwardPlan = [...this.state.backwardsSteps]
      .reverse()
      .map((step, index) => ({
        ...step,
        id: `forward-${index}`
      }));
    
    this.emit('plan-created', { 
      message: "Forward plan created",
      plan: this.state.forwardPlan
    });
  }
  
  /**
   * Private method to execute a single step
   * 
   * @param step The step to execute
   * @returns Promise that resolves with the step result
   */
  private async executeStep(step: Step): Promise<any> {
    for (const toolName of step.toolsNeeded) {
      if (!this.tools.has(toolName)) {
        throw new Error(`Missing required tool: ${toolName}`);
      }
    }
    
    await new Promise(resolve => setTimeout(resolve, 1000));
    
    const results = await Promise.all(
      step.toolsNeeded.map(async toolName => {
        const tool = this.tools.get(toolName);
        if (!tool) return null;
        
        try {
          return await tool.execute({ step });
        } catch (error) {
          this.emit('tool-error', { 
            message: `Error executing tool ${toolName}`,
            error
          });
          return null;
        }
      })
    );
    
    return {
      stepId: step.id,
      toolResults: results.filter(Boolean)
    };
  }
}
